### PR TITLE
broom for ml_kmeans()

### PR DESCRIPTION
**Pull request by samuelmacedo83**
I made a code for tidy(), augment() and glance(). I didn't create a testthat yet because I would like to know if the code is ok. Another thing, in spark the cluster number starts with 0, but in R starts with 1. I added 1 to cluster columns to be the same as R. 

Here is an example that you can run :

sc <- spark_connect(master = "local")
iris_tbl <- sdf_copy_to(sc, iris, name ="iris_tbl", overwrite = TRUE)

iris_tbl %>% 
  ml_kmeans(Species ~ ., k = 4, seed = 123) %>% 
  tidy()

iris_tbl %>% 
  ml_kmeans(Species ~ ., k = 4, seed = 123) %>% 
  augment()

iris_tbl %>% 
  ml_kmeans(Species ~ ., k = 4, seed = 123) %>% 
  augment(newdata = sdf_sample(iris_tbl,
                               fraction = 0.5))

iris_tbl %>% 
  ml_kmeans(Species ~ ., k = 4, seed = 123) %>% 
  glance()

### CODE REVIEW

**Comment by javierluraschi** [comment]
@samuelmacedo83 seems there are a couple documentation issues...

```
* checking R code for possible problems ... NOTE
augment.ml_model_kmeans: no visible binding for global variable 'label'
augment.ml_model_kmeans: no visible binding for global variable
  'features'
augment.ml_model_kmeans: no visible binding for global variable
  'prediction'
glance.ml_model_kmeans: no visible global function definition for
  'tibble'
Undefined global functions or variables:
  features label prediction tibble
* checking Rd files ... OK
* checking Rd metadata ... OK
* checking Rd line widths ... OK
* checking Rd cross-references ... OK
* checking for missing documentation entries ... OK
* checking for code/documentation mismatches ... OK
* checking Rd \usage sections ... WARNING
Duplicated \argument entries in documentation object 'ml_unsupervised_tidiers':
  'x'
Functions with \usage entries need to have the appropriate \alias
entries, and all their arguments documented.
The \usage entries must correspond to syntactically valid R code.

```

**Comment by javierluraschi** [comment]
@samuelmacedo83 Looking good!

Could we also add a test that compare against R models and an entry to the NEWS file? Something like:

```
## Broom
  
- Implemented `tidy()`, `augment()`, and `glance()` for `ml_kmeans()` models (@samuelmacedo83)
```

File: R/tidiers_ml_unsupervised_models.R [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+                                    ...){
+
+  # if the user doesn't provide a new data, this funcion will
+  # use the training set
+  if (is.null(newdata)){
+    newdata <- x$dataset
+  }
+
+  ml_predict(x, newdata) %>%
+    dplyr::select(-!!sym("label"), -!!sym("features")) %>%
```
Comment (kevinykuo) [inline R/tidiers_ml_unsupervised_models.R]: You can use the `.data` pronoun which is cleaner than unquoting symbols, e.g. `. %>% dplyr::select(-.data$label, -.data$features)`.

File: R/tidiers_ml_unsupervised_models.R [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    dplyr::mutate(prediction = !!sym("prediction") + 1) %>%
```
Comment (kevinykuo) [inline R/tidiers_ml_unsupervised_models.R]: Let's leave the cluster numbering as-is to be consistent with Spark.

File: R/tidiers_ml_unsupervised_models.R [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+#' @rdname ml_unsupervised_tidiers
+#' @export
+glance.ml_model_kmeans <- function(x,
+                                   ...) {
+
+  max.iter <- x$pipeline_model$stages[[2]]$param_map$max_iter
+  tol <- x$pipeline_model$stages[[2]]$param_map$tol
+  wssse <- x$cost
```
Comment (kevinykuo) [inline R/tidiers_ml_unsupervised_models.R]: Is this `totss` from `stats::kmeans()`?

File: R/tidiers_ml_unsupervised_models.R [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+                                 ...){
+  model <- x$model
+  center <- x$centers
+
+  stats <- c("cluster_sizes", "k")
+  statistics <- stats %>%
+    lapply(function(x) ml_summary(model, x, allow_null = TRUE))
+
+  cbind(center,
+        sizes = statistics[[1]],
```
Comment (kevinykuo) [inline R/tidiers_ml_unsupervised_models.R]: `sizes` -> `size`

File: R/tidiers_ml_unsupervised_models.R [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    dplyr::rename(.cluster = !!"prediction")
+
+}
+
+#' @rdname ml_unsupervised_tidiers
+#' @export
+glance.ml_model_kmeans <- function(x,
+                                   ...) {
+
+  max.iter <- x$pipeline_model$stages[[2]]$param_map$max_iter
```
Comment (kevinykuo) [inline R/tidiers_ml_unsupervised_models.R]: Maybe we leave out `max.iter` just in case people confuse it with `iter`?

Comment (samuelmacedo83) [reply R/tidiers_ml_unsupervised_models.R]: I was wondering the same thing and I agree with you. 
I didn't find if spark provides the number of iteration, I found just the max.iter and tolerance.
But if I remove max.iter and tol, glance() will have just one statistic, is it ok just print one statistic?

Comment (kevinykuo) [reply R/tidiers_ml_unsupervised_models.R]: I think that's fine.

File: R/tidiers_ml_unsupervised_models.R (hunk 2 above)
Comment (samuelmacedo83) [reply R/tidiers_ml_unsupervised_models.R]: Ok, do I have to change cluster column in tidy() too?

File: R/tidiers_ml_unsupervised_models.R (hunk 3 above)
Comment (samuelmacedo83) [reply R/tidiers_ml_unsupervised_models.R]: I guess is tot.withinss but I confess that I'm not quite sure.

**Comment by samuelmacedo83** [comment]
I updated my repository and now I have this problem

```
iris_tbl <- sdf_copy_to(sc, iris, name ="iris_tbl", overwrite = TRUE)

km_model <- iris_tbl %>% 
  ml_kmeans(Species ~ ., k = 4)
```
When I call cluster_sizes
```
ml_summary(km_model, metric = "cluster_sizes")
```
```
function() invoke(jobj, "clusterSizes")
<bytecode: 0x0000000011342130>
<environment: 0x000000001139c6e8>
```
or

```
km_model$summary$cluster_sizes()
```
```
 Error: java.lang.IllegalArgumentException: Object not found 551
	at sparklyr.StreamHandler.handleMethodCall(stream.scala:115)
	at sparklyr.StreamHandler.read(stream.scala:66)
	at sparklyr.BackendHandler.channelRead0(handler.scala:51)
	at sparklyr.BackendHandler.channelRead0(handler.scala:4)
	at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:105)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:102)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:310)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:284)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1359)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:935)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:138)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:645)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:580)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:497)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:459)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
	at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:138)
	at java.lang.Thread.run(Unknown Source) 
```

Any ideas?

**Comment by kevinykuo** [comment]
Might be a regression from my recent changes. Will take a look now.

**Comment by kevinykuo** [comment]
```
library(sparklyr)
sc <- spark_connect(master = "local")
iris_tbl <- sdf_copy_to(sc, iris, name ="iris_tbl", overwrite = TRUE)

km_model <- iris_tbl %>% 
  ml_kmeans(Species ~ ., k = 4)
km_model$summary$cluster_sizes()
# [1] 29 50 28 43
```

This was the breaking change from the PR merged today https://github.com/rstudio/sparklyr/pull/1622. The `km_model$summary$cluster_sizes()` should work though... could you try again?

**Comment by samuelmacedo83** [comment]
Yes, it's working :) 
I changed my code to call  x$summary$cluster_sizes(). 
I'll update it when a finish the test

**Comment by javierluraschi** [comment]
Easier to compare diff: https://github.com/rstudio/sparklyr/compare/master...samuelmacedo83:broom-kmeans

**Pull request merged**